File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    done = 0
    total_weight = 0
    time = 0
    
    bridge = deque([0 for _ in range(bridge_length)]) # 0으로 초기화된 다리
    trucks = deque(truck_weights) # 트럭들
    
    while done < len(truck_weights):
        done_truck = bridge.popleft()
        total_weight -= done_truck
        if done_truck != 0:
            done += 1
        
        if trucks:
            if trucks[0] + total_weight <= weight:
                truck = trucks.popleft()
                total_weight += truck
                bridge.append(truck)
            else:
                bridge.append(0)
        else:
            bridge.append(0)        
        time += 1
    
    return time

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.ArrayDeque;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        ArrayDeque<Integer> bridge = new ArrayDeque<>();
        for(int i = 0; i < bridge_length; i++){
            bridge.add(0);
        }
        
        ArrayDeque<Integer> trucks = new ArrayDeque<>();
        for(int truck : truck_weights){
            trucks.add(truck);
        }
        
        int done = 0;
        int time = 0;
        int currentWeight = 0;
        
        while(done < truck_weights.length){
            // 다리에서 빼내기
            int doneTruck = bridge.removeFirst();
            currentWeight -= doneTruck;
            if(doneTruck != 0) done++;
            
            if(!trucks.isEmpty()){
                // 트럭 있으면
                if(currentWeight + trucks.peekFirst() <= weight){
                    // 올릴 수 있으면
                    int nowTruck = trucks.removeFirst();
                    currentWeight += nowTruck;
                    bridge.addLast(nowTruck);
                }else{
                    // 올릴 수 없으면
                    bridge.addLast(0);
                }
            }else{
                bridge.addLast(0);
            }
            time++;
        }
        
        return time;
    }
}